sdk/core/allocator/alloc.h

```diff
@@ -130,53 +127,328 @@ static inline size_t leftshifted_val_msb(size_t val)
+	{
+		{
+			F(d)
+			} -> std::same_as<size_t>;
+	};
+
+	template<auto F, typename D>
+	concept Encoder = requires(size_t s)
+	{
+		{
+			F(s)
+			} -> std::same_as<D>;
+	};
+
+	/**
+	 * Equipped with a context for bounds, a reference to a displacement can be
+	 * a proxy for a pointer.
+	 */
+	template<typename T, typename D, bool Positive, auto Decode, auto Encode>
+	requires Decoder<Decode, D> && Encoder<Encode, D>
+	class Proxy
+	{
+		CHERI::Capability<void> ctx;
+		D                      &d;
+
+		__always_inline void set(T *p)
+		{
+			size_t diff =
+			  Positive ? ds::pointer::diff(ctx, p) : ds::pointer::diff(p, ctx);
+			d = Encode(diff);
+		}
+
+		public:
+		using Type = T;
+
+		__always_inline Proxy(void *c, D &r) : ctx(c), d(r) {}
+
+		__always_inline operator T *() const
+		{
+			size_t disp = Decode(d);
+
+			auto p = CHERI::Capability{ctx};
+			auto a = p.address();
+
+			if constexpr (Positive)
+			{
+				a += disp;
+			}
+			else
+			{
+				a -= disp;
+			}
+
+			return reinterpret_cast<T *>(a.ptr());
+		}
+
+		__always_inline T *operator->()
+		{
+			return *this;
+		}
+
+		__always_inline operator ptraddr_t()
+		{
+			return *this;
+		}
+
+		__always_inline Proxy &operator=(T *p)
+		{
+			set(p);
+			return *this;
+		}
+
+		__always_inline Proxy &operator=(Proxy const &p)
+		{
+			set(p);
+			return *this;
+		}
+
+		__always_inline bool operator==(Proxy const &p) const
+		{
+			return static_cast<T *>(*this) == static_cast<T *>(p);
+		}
+
+		__always_inline auto operator<=>(Proxy const &p) const
+		{
+			return static_cast<T *>(*this) <=> static_cast<T *>(p);
+		}
+	};
+
+	static_assert(ds::pointer::proxy::Proxies<
+	              Proxy<void, SmallSize, false, head2size, size2head>,
+	              void>);
+
+} // namespace displacement_proxy
+
 /**
- * When chunks are not in use, they are treated as nodes of either
+ * Every chunk, in use or not, includes a minimal header.  That is, this is a
+ * classic malloc, not something like a slab or sizeclass allocator or a
+ * "BIBOP"-inspired design.
+ *
+ * This header uses relative displacements to refer to the address-order
+ * predecessor and successor adjacent headers.  The details of encoding are
+ * encapsulated using the displacement_proxy::Proxy above and the
+ * cell_{next,prev} methods herein.
+ *
+ * Chunks are in one of three states:
+ *
+ *   - Allocated / "In Use" by the application
+ *
+ *       - Not indexed by any other structures in the MState
+ *
+ *   - Quarantined (until revocation scrubs inward pointers from the system)
+ *
+ *       - Collected in a quarantine ring using the MChunk::ring linkages
+ *
+ *   - Free for allocation
+ *
+ *       - Collected in a smallbin ring (using MChunk::ring) or in a treebin
+ *       ring (using either/both the TChunk and MChunk::ring links).
+ */
+struct __packed __aligned(MallocAlignment)
+MChunkHeader
+{
+	/**
+	 * Compressed size of the predecessor chunk.  See cell_prev().
+	 */
+	SmallSize sprev;
```

Comment:
Very nit: `sprev` and `shead` used to contain both size and other metadata bits. Now they are just sizes, we can give them better names like `prevSize` and `currSize` if it doesn't require too many changes.